add product Get Validator add product Get Serializer

gostishev opened the pull request.

Commit:
add product Get Validator add product Get Serializer

Commit:
feat(Product):finish routers in ProductController

add routers PUT, PATCH, GET/{id}, DELETE in ProductController

Commit:
feat(order):add entitys BillingType Order OrderItem

insert into billing_type from migration by 'hard', add router POST in OrderController

Commit:
feat(order):add routers GET/{id} GET DELETE

In OrderController fix POST router(orderGetSerializer and orderItemOutputDataTransform)
In Delete router used orphanRemoval=true, In post router you can see cascade={persist}

ossman-d commented on migrations/Version20220808073658.php:
```diff
@@ -0,0 +1,38 @@
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
```
Убрать все мусорные комменты

ossman-d commented on migrations/Version20220808090836.php:
```diff
@@ -0,0 +1,36 @@
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
```
Убрать все мусорные комменты и закомменченый код

ossman-d commented on migrations/Version20220808125131.php:
```diff
@@ -0,0 +1,40 @@
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
```
Убрать все мусорные комменты и закомменченый код

ossman-d commented on migrations/Version20220808125656.php:
```diff
@@ -0,0 +1,36 @@
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
```
Убрать все мусорные комменты и закомменченый код

ossman-d commented on migrations/Version20220808132414.php:
```diff
@@ -0,0 +1,42 @@
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
```
Убрать все мусорные комменты и закомменченый код

ossman-d commented on migrations/Version20220808142949.php:
```diff
@@ -0,0 +1,42 @@
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
```
Убрать все мусорные комменты и закомменченый код

ossman-d commented on migrations/Version20220810081622.php:
```diff
@@ -0,0 +1,49 @@
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
```
Убрать все мусорные комменты и закомменченый код

ossman-d commented on migrations/Version20220814132223.php:
```diff
@@ -0,0 +1,46 @@
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
```
Убрать все мусорные комменты и закомменченый код

ossman-d commented on src/Controller/CategoryController.php:
```diff
@@ -47,7 +47,7 @@ public function getCategories(ManagerRegistry $doctrine, Request $request): Json
             $order = $request->query->get('order');
```
Необходимо разделить контроллер для лучшей читабельности.
Завести следующую структуру: `src/Controller/Category/`
внутри директории классы:
- GetController
- ShowController
- RemoveController
- AddController
- ReplaceController
- EditController

В новых контроллерах код переносится в метод `__invoke`. Аннотация с роутом пишется на весь класс.

Метод `transformJsonBody` вынести в отдельный класс src/Helper/JsonBody.php.

Commit:
fix(migrations):clean up migrations

drop garbage comments and commented code

Commit:
fix(category):Move the routers to separate classes

Commit:
feat(order):add router PUT

ossman-d commented on src/Helper/TransformJsonBody.php:
```diff
@@ -0,0 +1,21 @@
+
+use Symfony\Component\HttpFoundation\Request;
+
+class TransformJsonBody
```
переименуй класс в `JsonBody`

ossman-d commented on src/Helper/TransformJsonBody.php:
```diff
@@ -0,0 +1,21 @@
+
+class TransformJsonBody
+{
+    public function transformJsonBody(Request $request)
```
переименуй метод в `transform`

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,60 @@
+    public function __invoke(ManagerRegistry $doctrine, Request $request, ValidatorInterface $validator): JsonResponse
+    {
+        try {
+            $request = (new TransformJsonBody())->transformJsonBody($request);
```
перезаписывать `$request` не нужно. Он передаётся по ссылке и все изменения в рамках этого вызова апи сохраняются внутри класса `Request`.

ossman-d replied:
https://symfony.com/bundles/SensioFrameworkExtraBundle/current/annotations/converters.html - вот инструмент, для преобразования входящих параметров. Попробуй использовать его

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,60 @@
+            $violations = $validator->validate($dto);
+
+            if (0 !== count($violations)) {
+                throw  new CustomErrorException("", 422, null, $violations->getIterator());
```
ctrl+alt+L - убирает лишние пробелы (`throw__new`)

ossman-d replied:
всё ещё не убрал лишний пробел

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,60 @@
+            $category = new Category();
+            $category->setName($request->get('name'));
+            $category->setSort($request->request->get('sort'));
+            $entityManager = $doctrine->getManager();
```
Если ты вызываешь `ManagerRegistry` через инъекцию зависимостей, почему сразу не вызвать там `EntityManagerInterface`?
 Более простым языком - замени `ManagerRegistry` на входе в метод и `$entityManager = $doctrine->getManager();` на
`EntityManagerInterface $entityManager` на входе в метод

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,60 @@
+            $entityManager->persist($category);
+            $entityManager->flush();
+
+            $classMetadataFactory = new ClassMetadataFactory(new AnnotationLoader(new AnnotationReader()));
```
С 49-й по 52-ю строку можно целиком вынести на это событие:
https://symfony.com/doc/current/reference/events.html#kernel-response.
В `Response` отдавать не сериализованные данные и сериализовать их централизованно в одном месте.

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,60 @@
+            $serializer = new Serializer([$normalizer]);
+            $data = $serializer->normalize($category, null, ['groups' => 'group1']);
+
+            return (new ResponseJson())->response($data);
```
От класса `ResponseJson` можно избавиться, т.к. он не несёт ничего в себе и является просто обёрткой на встроенным `JsonResponse`. Твои контроллеры вполне валидно могут возвращаться `JsonResponse`, как более частную версию `Response`.

Commit:
fix(category)fix routrers ofcategory entity

add  src/EventListener/JsonBodySubscriber.php
add src/EventListener/JsonResponseSubscriber.php
on event kernel.controller and on event kernel.view accordingly

Commit:
fix(product):move the routers to separate classes

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,45 @@
+            }
+
+            $category = new Category();
+            $category->setName($request->get('name'));
```
почему на следующей строке `$request->request->get()`, а на этой `$request->get()`?

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,45 @@
+            $entityManager->flush();
+
+            return $category;
+
```
лишняя пустая строка. Следи за этим в своём коде

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,45 @@
+
+            return $category;
+
+        } catch (CustomErrorException $e) {
```
пора избавляться от try-catch в контроллере и обрабатывать ошибки в соответствующем событии:
https://symfony.com/doc/current/reference/events.html#kernel-exception
Примерный гайд, как обрабатывать ошибки:
https://inviqa.com/blog/guide-custom-error-handling-symfony

ossman-d commented on src/EventListener/JsonBodySubscriber.php:
```diff
@@ -0,0 +1,30 @@
+            return $request;
+        }
+
+        $request->request->replace($data);
```
обязательно нужна пустая строка перед между строкой кода и `return`. Это стандарт PSR

ossman-d commented on src/Helper/CountEntityProduct.php:
```diff
@@ -0,0 +1,30 @@
+use App\Entity\Product;
+use Symfony\Bridge\Doctrine\ManagerRegistry;
+
+class CountEntityProduct
```
Плохой хэлпер. У тебя для этого есть `Repository` для каждой сущности (`Entity`). Вынеси этот код туда.

ossman-d replied:
Если функционал этого класса уже вынесен в `ProductRepository`, то класс нужно удалить

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,103 @@
+                );
+            }
+
+            $categoryIdNowThereIs = $product->getCategory()->getId();
```
Непонятное имя переменной. Так и не смог понять, какой что несёт в себе это название.
Назови его просто `$productCategoryId`

ossman-d commented on src/Controller/Category/EditController.php:
```diff
@@ -0,0 +1,75 @@
+
+            $violations = new ConstraintViolationList();
+
+            if ($request->request->has('name')) {
```
всю валидацию до 51-й строки нужно переносить в `CategoryEditInputDTO`, в котором будут не такие строгие правила по поводу отсутствующих значений

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,45 @@
+        try {
+            $dto = new CategoryInputDTO($request->get('name'), $request->get('sort'));
+            /** @var   ConstraintViolationList $violations */
+            $violations = $validator->validate($dto);
```
всю конструкцию 
```
$violations = $validator->validate($dto);
if (0 !== count($violations)) {
                throw  new CustomErrorException("", 422, null, $violations->getIterator());
}
```
надо бы вынести в отдельный класс, типа `AbtractValidator`, который будет вызывать во всех контроллерах, где есть `input DTO`.
Ниже есть коммент про try-catch. При условии реализации этих двух пунктов, код всех контроллеров на запись-сохранение-обновление будет намного чище.

ossman-d commented on src/Controller/Category/GetController.php:
```diff
@@ -0,0 +1,40 @@
+            $order = $request->query->has('order') ? $request->query->get('order') : 'ASC';
+
+            if (!in_array($order, ['ASC', 'DESC'])) {
+                throw new NotFoundHttpException();
```
Почему здесь `NotFoundHttpException`?

ossman-d commented on src/Controller/Category/ReplaceController.php:
```diff
@@ -0,0 +1,57 @@
+
+            $dto = new CategoryInputDTO($request->get('name'), $request->get('sort'));
+            /** @var   ConstraintViolationList $violations */
+            $violations = $validator->validate($dto);
```
опять же, для всей этой конструкции будет `AbstractValidator`, о котором я писал раньше

ossman-d commented on src/Controller/Category/GetController.php:
```diff
@@ -0,0 +1,40 @@
+        } catch (NotFoundHttpException $e) {
+            $data = [
+                'status' => 422,
+                'errors' => "Query parameter 'order' not equal to 'ASC' or 'DESC'",
```
Почему текст при обработке эксепшена написан хардкодом?
Если выносить обработку эксепшенов на событие, как будешь определять, какой текст показывать в респонсе?

ossman-d commented on src/Controller/Category/RemoveController.php:
```diff
@@ -0,0 +1,43 @@
+
+            if (!$category) {
+                throw new NotFoundHttpException(
+                    "Category not found for id: $id", null, 404
```
Добавил текст в эксепшен, а ниже всё равно вывел такой же текст хардкодом. Зачем?

ossman-d commented on src/Controller/Category/RemoveController.php:
```diff
@@ -0,0 +1,43 @@
+        } catch (NotFoundHttpException $e) {
+            $data = [
+                'status' => 404,
+                'errors' => "Category not found for id: $id",
```
Почему текст при обработке эксепшена написан хардкодом?
Если выносить обработку эксепшенов на событие, как будешь определять, какой текст показывать в респонсе?

ossman-d commented on src/Controller/Category/RemoveController.php:
```diff
@@ -0,0 +1,43 @@
+
+        } catch (NotFoundHttpException $e) {
+            $data = [
+                'status' => 404,
```
Ты в эксепшене даже статус-код передаёшь на 26-й строке

ossman-d commented on src/Controller/Category/ReplaceController.php:
```diff
@@ -0,0 +1,57 @@
+
+            if (!$category) {
+                throw new NotFoundHttpException(
+                    "Category not found for id: $id", null, 404
```
Уже писал об этом

ossman-d commented on src/Controller/Category/ReplaceController.php:
```diff
@@ -0,0 +1,57 @@
+        } catch (NotFoundHttpException $e) {
+            $data = [
+                'status' => 404,
+                'errors' => "Category not found for id: $id",
```
Уже писал об этом выше

ossman-d commented on src/Controller/Category/ReplaceController.php:
```diff
@@ -0,0 +1,57 @@
+
+        } catch (NotFoundHttpException $e) {
+            $data = [
+                'status' => 404,
```
Уже писал об этом выше

ossman-d commented on src/Controller/Category/ShowController.php:
```diff
@@ -0,0 +1,40 @@
+
+        } catch (NotFoundHttpException $e) {
+            $data = [
+                'status' => 404,
```
уже писал об этом выше

ossman-d commented on src/Controller/Category/ShowController.php:
```diff
@@ -0,0 +1,40 @@
+        } catch (NotFoundHttpException $e) {
+            $data = [
+                'status' => 404,
+                'errors' => "Category not found for id: $id",
```
уже писал об этом выше

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,103 @@
+                throw  new CustomErrorException("", 422, null, $violations->getIterator());
+            }
+
+            $product->SetName($result["name"]);
```
`SetName` - не по PSR

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,103 @@
+                'price' => new PriceProductConstraint(),
+                'category_id' => new CategoryProductConstraint(),
+            ];
+            foreach ($properties as $key => $value) {
```
Неплохая реализация. Интересный ход мыслей, чтобы автоматизировать рутину.
Попробуй, всё-таки, перенести constraint-ы в `ProductEditInputDTO` и провадилировать его абстракным валидатором, о котором я писал раньше.

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,103 @@
+            }
+            $product->setCategory($category);
+            $entityManager->flush();
+            $productRepo[] = $product;
```
зачем это?

Commit:
fix(category):-try_catch +NotPassedClass in PATCH +ExceptionSubscriber

Commit:
fix(category):+ValidatorInputDTO for validation all inputDTO

Commit:
fix(product):fix AddController,EditController

Commit:
fix(product):fix GET DELETE PUT GET/id routers

Commit:
fix(order):move the routers to separete classes

Commit:
fix(order):add validation query parameter with the help OrderGetInputDTO

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,33 @@
+ * @Route("/category", name="category_add", methods={"POST"})
+ */
+class AddController extends AbstractController
+//"http://localhost:8082/category"
```
таким комментам не место в коде. Сохраняй себе отдельные пути для каждого http-метод в insomnia

ossman-d commented on src/Controller/Category/EditController.php:
```diff
@@ -0,0 +1,47 @@
+ * @Route("/category/{id}", name="category_edit", methods={"PATCH"})
+ */
+class EditController extends AbstractController
+//"http://localhost:8082/category/id"
```
удалить

ossman-d commented on src/Controller/Category/GetController.php:
```diff
@@ -0,0 +1,29 @@
+ * @Route("/category", name="category_get", methods={"GET"})
+ */
+class GetController extends AbstractController
+//"http://localhost:8082/category?order=ASC"
```
удалить

ossman-d commented on src/Controller/Category/RemoveController.php:
```diff
@@ -0,0 +1,31 @@
+ * @Route("/category/{id}", name= "category_remove", methods={"DELETE"})
+ */
+class RemoveController extends AbstractController
+//"http://localhost:8082/category/id"
```
удалить

ossman-d commented on src/EventListener/ExceptionSubscriber.php:
```diff
@@ -0,0 +1,39 @@
+//        dump($exception->getCode());
+        if ($exception instanceof NotFoundHttpException) {
+            $response = new JsonResponse($exception->getMessage(), $exception->getCode());
+            $event->setResponse($response);
```
этот код и код на строке 35 дублируется. Можно оставить одну такую строку после обоих `if`

ossman-d commented on src/Helper/ValidatorInputDTO.php:
```diff
@@ -0,0 +1,21 @@
+{
+    public function validateInput(ValidatorInterface $validator, $dto)
+    {
+        /** @var   ConstraintViolationList $violations */
```
три пробела вместо одного в аннотации

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,33 @@
+        (new ValidatorInputDTO())->validateInput($validator, $dto);
+
+        $category = new Category();
+        $category->setName($request->request->get('name'));
```
не критично, но лучше заполнять `Entity` из `DTO`: `$dto->name` вместо `$request->request->get('name')`. На следующей строке аналогично.
При записи свойств в DTO у тебя может произойти какая-нибудь трансформация данных. Например `Unix timestamp` будет преобразовываться в `DateTimeImmutable`. Тогда наколешься на записи напрямую из тела запроса в сущность.

ossman-d commented on src/Controller/Category/AddController.php:
```diff
@@ -0,0 +1,33 @@
+
+        $category = new Category();
+        $category->setName($request->request->get('name'));
+        $category->setSort($request->request->get('sort'));
```
аналогично предыдущей строке

ossman-d commented on src/DTO/CategoryGetInputDTO.php:
```diff
@@ -0,0 +1,23 @@
+use App\Validator as CustomAssert;
+use Symfony\Component\Validator\Constraints\ExpressionSyntax;
+
+class CategoryGetInputDTO
```
Я бы переименовал в `CategoryListQueryDTO`, т.к. `Input` в названии предполагает, что ты записываешь в эту DTO входящие данные сущности.
А `Query` - будет указывать на то, что DTO хранит в себе параметры запроса.

ossman-d commented on src/Controller/Category/RemoveController.php:
```diff
@@ -0,0 +1,31 @@
+                "Category not found for id: $id", null, 404
+            );
+        }
+            $entityManager->remove($category);
```
неправильные отступы (4 лишних пробела слева от кода в этой строке)

ossman-d commented on src/Controller/Category/RemoveController.php:
```diff
@@ -0,0 +1,31 @@
+            );
+        }
+            $entityManager->remove($category);
+            $entityManager->flush();
```
неправильные отступы (4 лишних пробела слева от кода в этой строке)

ossman-d commented on src/Controller/Category/ReplaceController.php:
```diff
@@ -0,0 +1,39 @@
+ * @Route("/category/{id}", name="category_replace", methods={"PUT"})
+ */
+class ReplaceController extends AbstractController
+//"http://localhost:8082/category/id"
```
удалить

ossman-d commented on src/Controller/Category/ReplaceController.php:
```diff
@@ -0,0 +1,39 @@
+        $dto = new CategoryInputDTO($request->request->get('name'), $request->request->get('sort'));
+        (new ValidatorInputDTO())->validateInput($validator, $dto);
+
+        $category->setName($request->request->get('name'));
```
В `AddController` уже описал, почему лучше не записывать данные из тела реквеста в сущность.

ossman-d commented on src/Controller/Category/ReplaceController.php:
```diff
@@ -0,0 +1,39 @@
+        (new ValidatorInputDTO())->validateInput($validator, $dto);
+
+        $category->setName($request->request->get('name'));
+        $category->setSort($request->request->get('sort'));
```
Аналогично предыдущей строке

ossman-d commented on src/Controller/Category/ShowController.php:
```diff
@@ -0,0 +1,29 @@
+{
+    public function __invoke(ManagerRegistry $doctrine, int $id): mixed
+    {
+            $category = $doctrine->getRepository(Category::class)->find($id);
```
лишние пробелы во всё коде метода (неправильные отступы)

ossman-d commented on src/Controller/Category/ShowController.php:
```diff
@@ -0,0 +1,29 @@
+ * @Route("/category/{id}", name="category_show", methods={"GET"})
+ */
+class ShowController extends AbstractController
+//"http://localhost:8082/category/id"
```
удалить

ossman-d commented on src/Controller/Product/AddController.php:
```diff
@@ -0,0 +1,55 @@
+            );
+        }
+
+        $product->setName($request->request->get('name'));
```
Лучше заполнять `Entity` из `DTO`: `$dto->name` вместо `$request->request->get('name')`. Для всех свойств сущности ниже, кроме `->setCategory($category)`, тот же самый совет.

ossman-d commented on src/Controller/Product/AddController.php:
```diff
@@ -0,0 +1,55 @@
+ * @Route("/product", name="product_add", methods={"POST"})
+ */
+class AddController extends AbstractController
+//"http://localhost:8082/product"
```
удалить

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,65 @@
+ * @Route("/product/{id}", name="product_edit", methods={"PATCH"})
+ */
+class EditController extends AbstractController
+//"http://localhost:8082/category/id"
```
удалить

ossman-d commented on src/Helper/GetSerializer.php:
```diff
@@ -0,0 +1,30 @@
+
+use App\DTO\ProductOutputDTO;
+
+class GetSerializer
```
Неправильное именование классов и из-за этого происходит путаница в понимании, что он делает.
**Совет**: назвать класс `ProductEntityOutputDtoTransformer`. Вынести его в отдельную директорию `src/DataTransformer`.

ossman-d commented on src/Helper/GetSerializer.php:
```diff
@@ -0,0 +1,30 @@
+
+class GetSerializer
+{
+    public function outputDtoSerializer($product): ProductOutputDTO
```
Вместе с неправильным именованием класса также неправильно назван метод, что вносит путаницу при чтении кода.
Метод не содержит в своём названии глагола, что является грубой ошибкой при выборе имени.
Поскольку выше я описал подробное название класса, рекомендую назвать метод просто `transform`.
Также не указан тип входного параметра `$product`.

ossman-d commented on src/Helper/GetSerializer.php:
```diff
@@ -0,0 +1,30 @@
+        );
+    }
+
+    public function arrOutputDtoSerializer($productsRepo):mixed
```
Рекомендую переименовать в `arrayTransform`.
Входной параметр `$productsRepo` лучше назвать `$productArray`. Не указан тип входного параметра.
Возвращаемый тип у тебя здесь не `mixed`, а `array`.

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,65 @@
+        $categoryIdRepo = $product->getCategory()->getId();
+
+        $inputDto = new ProductInputDTO(
+            $name = $request->request->has('name') ? $request->request->get('name') : $product->getName(),
```
Лучше заполнять `Entity` из `DTO`: $dto->name вместо $request->request->get('name'). Для всех свойств сущности ниже, тот же самый совет.
Писал об этом уже и раньше.

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,65 @@
+                "Product  not found for id: $id", null, 404
+            );
+        }
+        $categoryIdRepo = $product->getCategory()->getId();
```
Почему в названии присутствует слово `Repo`? Какую смысловую нагрузку оно несёт? Если никакой, то имени `$productCategoryId` будет достаточно

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,65 @@
+        $inputDto = new ProductInputDTO(
+            $name = $request->request->has('name') ? $request->request->get('name') : $product->getName(),
+            $price = $request->request->has('price') ? $request->request->get('price') : $product->getPrice(),
+            $categoryId = $request->request->has('category_id') ? $request->request->get('category_id') : $categoryIdRepo,
```
Тут логика ясна. Но присвоение на 45-й строке - нет.
При именовании переменных по моему совету получишь следующий код:
```
$categoryId = $request->request->has('category_id') ? $request->request->get('category_id') : $productCategoryId,
```
Продолжение на строке 48

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,65 @@
+        $product->setPrice($price);
+        $product->setCreatedAt(new \DateTimeImmutable());
+        $product->setDescription($description);
+        $categoryIdRequest = $categoryId;
```
вот это место, как по мне, не нужно. Потому что в `$categoryIdRequest` попадёт при переданном параметре `category_id` - значение этого парамерта.
При не переданном - `$productCategoryId` (бывший `$categoryIdRepo`)

ossman-d commented on src/Controller/Product/EditController.php:
```diff
@@ -0,0 +1,65 @@
+        $categoryIdRequest = $categoryId;
+
+
+        if ($categoryIdRepo === $categoryIdRequest) {
```
в результате, если внимательно прочесть комменты на 36-й и 45-й строках, это сравнение можно преобразовать в следующее:
```
if ($productCategoryId === $categoryId) { //$productCategoryId - это бывший $categoryIdRepo
```

ossman-d commented on src/Controller/Product/GetController.php:
```diff
@@ -0,0 +1,58 @@
+//  "http://localhost:8082/product?orderType=ASC&orderBy=id&filterBy=category&filterValue=26&pageSize=5&page=1"
+{
+    public function __invoke(EntityManagerInterface $entityManager, Request $request, ValidatorInterface $validator): mixed
+
```
лишняя пустая строка

ossman-d commented on src/Controller/Product/GetController.php:
```diff
@@ -0,0 +1,58 @@
+class GetController extends AbstractController
+//  "http://localhost:8082/product?orderType=ASC&orderBy=id&filterBy=category&filterValue=26&pageSize=5&page=1"
+{
+    public function __invoke(EntityManagerInterface $entityManager, Request $request, ValidatorInterface $validator): mixed
```
возвращаемый тип - `array`, а не `mixed`

ossman-d commented on src/Controller/Product/GetController.php:
```diff
@@ -0,0 +1,58 @@
+ * @Route("/product", name="product_get", methods={"GET"})
+ */
+class GetController extends AbstractController
+//  "http://localhost:8082/product?orderType=ASC&orderBy=id&filterBy=category&filterValue=26&pageSize=5&page=1"
```
удалить коммент. Переноси все эти штуки в insomnia

ossman-d commented on src/Helper/ValidatorPaginator.php:
```diff
@@ -0,0 +1,69 @@
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+use App\Helper\NotPassedClass;
+
+class ValidatorPaginator
```
Ты запихнул 3 ответственности в 1 класс:
1. расчёт параметров пагинации;
2. валидацию этих параметров;
3. валидацию параметра `filterValue`.
Делать так нельзя. Выноси валидацию в для query-параметров пагинации и `filterValue` и валидируй там. А калькуляцию оставь здесь.

Переименуй класс в `PaginationCalculator`.

ossman-d commented on src/Helper/ValidatorPaginator.php:
```diff
@@ -0,0 +1,69 @@
+
+class ValidatorPaginator
+{
+    public function offsetCalculateValidate(int $page, int $pageSize, int $entityCount, ValidatorInterface $validator): array|JsonResponse
```
переименуй в `offsetCalculate`.
Не вижу, чтобы этот метод где-то возвращал `JsonResponse`. Проверь возвращаемые значения

ossman-d commented on src/Helper/ValidatorPaginator.php:
```diff
@@ -0,0 +1,69 @@
+
+    }
+
+    public function validateFilterValue(EntityManagerInterface $entityManager, $filterValue): mixed
```
В комменте выше написал, что этой валидации в этом классе быть не должно.
У тебя здесь ещё и какой-то подсчёт сущностей идёт в результат. Валидация - отдельно. Калькуляция - отдельно. Подсчёт количества сущностей (выборка из БД) - отдельно.
Нужно распутывать этот клубок, т.к. этот класс явно плохо пахнет и оставляет ощущение грязных рук.

ossman-d commented on src/Controller/Product/GetController.php:
```diff
@@ -0,0 +1,58 @@
+            $validatorPaginator = new ValidatorPaginator();
+            $entityCount = $validatorPaginator->validateFilterValue($entityManager, $filterValue);
+
+            [$pageSize, $offset] = $validatorPaginator->offsetCalculateValidate($page, $pageSize, $entityCount, $validator);
```
этот вызов дублируется на строке 52, в чём никакой необходимости нет.
Сценарий расчёта параметров пагинации у тебя расчитывается при каждого запросе в этот контроллер. Можешь смело выносить его на несколько строк выше, чтобы он вызывался в одном месте.

ossman-d commented on src/Controller/Product/RemoveController.php:
```diff
@@ -0,0 +1,32 @@
+ * @Route("/product/{id}", name= "product_remove", methods={"DELETE"})
+ */
+class RemoveController extends AbstractController
+//"http://localhost:8082/product/id"
```
удалить коммент. Сохрани себе в isnomnia

ossman-d commented on src/Controller/Product/RemoveController.php:
```diff
@@ -0,0 +1,32 @@
+                "Product not found for id: $id", null, 404
+            );
+        }
+            $entityManager->remove($product);
```
лишние 4 пробела слева. Неправильный отступ

ossman-d commented on src/Controller/Product/RemoveController.php:
```diff
@@ -0,0 +1,32 @@
+            );
+        }
+            $entityManager->remove($product);
+            $entityManager->flush();
```
лишние 4 пробела слева. Неправильный отступ

ossman-d commented on src/Controller/Product/ReplaceController.php:
```diff
@@ -0,0 +1,63 @@
+ * @Route("/product/{id}", name="product_replace", methods={"PUT"})
+ */
+class ReplaceController extends AbstractController
+//"http://localhost:8082/product/id"
```
удалить коммент. Переноси все эти штуки в insomnia

ossman-d commented on src/Controller/Product/GetController.php:
```diff
@@ -0,0 +1,58 @@
+        (new ValidatorInputDTO())->validateInput($validator, $dto);
+
+        if ($request->query->has('filterBy')) {
+            $validatorPaginator = new ValidatorPaginator();
```
В этом классе допущено много ошибок. Сначала исправь его, а код ниже, скорее всего, после этого придётся переписать.

ossman-d commented on src/Controller/Product/GetController.php:
```diff
@@ -0,0 +1,58 @@
+
+            [$pageSize, $offset] = $validatorPaginator->offsetCalculateValidate($page, $pageSize, $entityCount, $validator);
+
+            $productsRepo = $entityManager->getRepository(Product::class)->findBy([$filterBy => $filterValue], [$orderBy => $orderType], $pageSize, $offset);
```
Тут и на 54-й строке вызов отличается только массивом фильтрации. Внутри условной логики сформируй отдельный массив `$filterParamsArray` (пустой массив или с параметрами, в зависимости от логики). А в районе 54-й строки подставь его в метод `findBy`. Не придётся 2 `return` в контроллере делать.

ossman-d commented on src/Controller/Product/RemoveController.php:
```diff
@@ -0,0 +1,32 @@
+            $entityManager->remove($product);
+            $entityManager->flush();
+
+        return (new GetSerializer())->outputDtoSerializer($product);
```
не нужен тут объект `Product` и сериализация не нужна. Пустого респонса и статуса 200 хватит

ossman-d commented on src/Controller/Product/ReplaceController.php:
```diff
@@ -0,0 +1,63 @@
+        $product->setName($request->request->get('name'));
+        $product->setDescription($description);
+        $product->setPrice($request->request->get('price'));
+        $product->setCreatedAt(new \DateTimeImmutable());
```
это лишняя логика. Дата создания в БД должна быть одна и не должна меняться. Можно удалить.
